pycozmo/brain.py

```diff
@@ -0,0 +1,138 @@
+"""
+
+Brain class.
+
+"""
+
+from threading import Thread
+from queue import Queue, Empty
+import time
+
+from . import logger
+from . import event
+from . import emotions
+from . import behavior
+from . import activity
+from . import anim
+
+
+__all__ = [
+    "Brain"
+]
+
+
+class Brain:
+    """ Cozmo robot brain class. """
+
+    def __init__(self, cli):
+        super().__init__()
+
+        self.activities = activity.load_activities()
+        self.behaviors = behavior.load_behaviors()
+        self.animation_groups = anim.load_animation_groups()    # TODO: Move to Client?
+        self.emotion_types = {emotion.name: 0.0 for emotion in emotions.EmotionType}
+        self.emotion_events = emotions.load_emotion_events()
+
+        self.cli = cli
+        self.cli.load_anims("...")
+        self.cli.add_handler(event.EvtCliffDetectedChange, self.on_cliff_detected)
+        # TODO: ...
+
+        # Reaction trigger queue
+        self.reaction_queue = Queue()
+
+        self.stop_flag = False
+        self.reaction_thread = Thread(daemon=True, name="ReactionThread", target=self.reaction_thread_run)
+        self.heartbeat_thread = Thread(daemon=True, name="HeartbeatThread", target=self.heartbeat_thread_run)
+
+    def start(self):
+        # Connect to robot
+        self.reaction_thread.start()
+        self.heartbeat_thread.start()
+
+        # TODO: Enable camera
+
+    def stop(self):
+        # Disconnect from robot
+        self.stop_flag = True
+        self.heartbeat_thread.join()
+        self.reaction_thread.join()
+
+    def on_cliff_detected(self, cli, state: bool) -> None:
+        if state:
+            self.post_reaction("CliffDetected")
+
+    def on_camera_image(self, cli, new_im) -> None:
+        """ Process images, coming from the robot camera. """
+        # TODO: motion detection
+        # self.process_reaction_trigger("UnexpectedMovement")
+        # TODO: face detection
+        # self.process_reaction_trigger("FacePositionUpdate")?
+        # TODO: pet detection
+        # self.process_reaction_trigger("PetInitialDetection")
+        # TODO: laser detection
+        # TODO: cube marker detection
+        # TODO: facial expression estimation
+        # TODO: smile amount detection
+        # TODO: blink amount detection
+        # TODO: gaze detection?
+        # TODO: image quality check
+        pass
+
+    def post_reaction(self, reaction_trigger: str) -> None:
+        """ Post a reaction trigger to the reaction trigger queue. """
+        self.reaction_queue.put(reaction_trigger)
+
+    def reaction_thread_run(self) -> None:
+        """ Reaction thread loop. Reaction trigger queue processing. """
+        while not self.stop_flag:
+            try:
+                reaction_trigger = self.reaction_queue.get(timeout=0.05)
```

Comment:
I'm liking a lot how useful queues are in python to manage multithreaded applications but I am wondering, is it necessary here? we could skip the reaction thread and process the reactions directly with the necessary callbacks and a couple of locks, which I guess we would need anyways.

Reply:
We could. I am open to removing it at some point if it proves to do not provide value.

The advantage of a dedicated thread is that it is clear that there is one stream of "reaction triggers" and processing them will always happen in the context of the "reaction thread". Without a queue and a dedicated thread, reaction triggers will be arising in the contexts of different threads (the connection thread, timers, maybe some external inputs) and this will lead to:
- potentially processing several in parallel
- slowing down other threads that have other things to do

Reply:
Okay, I understand, that way we offload the other threads from whatever processing is needed here. I like it better now :)